pkg/sql/temporary_tables_deletion.go

```diff
@@ -0,0 +1,226 @@
+	}
+	dbIDs, err := GetAllDatabaseDescriptorIDs(ctx, p.txn)
+	if err != nil {
+		return err
+	}
+	for _, dbID := range dbIDs {
+		schemaNames, err := p.GetAllSchemaNames(ctx, dbID, p.txn)
+		if err != nil {
+			return err
+		}
+		for _, scName := range schemaNames {
+			isTempSchema, sessionID, err := temporarySchemaSessionID(scName)
+			if err != nil {
+				return err
+			}
+			if isTempSchema && !activeSessions[sessionID.Uint128] {
+				if err := cleanupSessionTempObjects(ctx, p, sessionID); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// OnFailOrCancel is part of the jobs.Resumer interface.
+func (r *deleteTempTablesResumer) OnFailOrCancel(context.Context, *client.Txn) error {
+	return nil
+}
+
+// OnSuccess is part of the jobs.Resumer interface.
+func (r *deleteTempTablesResumer) OnSuccess(ctx context.Context, _ *client.Txn) error { return nil }
+
+// OnTerminal is part of the jobs.Resumer interface.
+func (r *deleteTempTablesResumer) OnTerminal(context.Context, jobs.Status, chan<- tree.Datums) {}
+
+type concurrentDeleteTempTablesError struct{}
+
+var _ error = concurrentDeleteTempTablesError{}
+
+func (concurrentDeleteTempTablesError) Error() string {
+	return "another delete temp tables job is already running"
+}
+
+// ConcurrentDeleteTempTablesError is reported when two delete temp tables jobs
+// are issued concurrently. This is a sentinel error.
+var ConcurrentDeleteTempTablesError error = concurrentDeleteTempTablesError{}
+
+func checkRunningTempTablesJob(ctx context.Context, job *jobs.Job, p *planner) error {
+	var jobID int64
+	if job != nil {
+		jobID = *job.ID()
+	}
+	const stmt = `SELECT id, payload FROM system.jobs WHERE status IN ($1, $2, $3) ORDER BY created`
+
+	rows, err := p.ExecCfg().InternalExecutor.Query(
+		ctx,
+		"get-jobs",
+		nil, /* txn */
+		stmt,
+		jobs.StatusPending,
+		jobs.StatusRunning,
+		jobs.StatusPaused,
+	)
+	if err != nil {
+		return err
+	}
+
+	for _, row := range rows {
+		payload, err := jobs.UnmarshalPayload(row[1])
+		if err != nil {
+			return err
+		}
+
+		if payload.Type() == jobspb.TypeDeleteTempTables {
+			id := (*int64)(row[0].(*tree.DInt))
+			if *id == jobID {
+				break
+			}
+
+			// This is not the first DeleteTempTables job running. This job should fail
+			// so that the earlier job can succeed.
+			return ConcurrentDeleteTempTablesError
+		}
+	}
+	return nil
+}
+
+func (p *planner) startJob(ctx context.Context, resultsCh chan<- tree.Datums) error {
+	if err := checkRunningTempTablesJob(ctx, nil /* job */, p); err != nil {
+		return err
+	}
+	record := jobs.Record{
+		Description: "delete temp tables of inactive sessions",
+		Username:    p.User(),
+		Details:     jobspb.DeleteTempTablesDetails{},
+		Progress:    jobspb.DeleteTempTablesProgress{},
+	}
+
+	job, errCh, err := p.ExecCfg().JobRegistry.StartJob(ctx, resultsCh, record)
+	if err != nil {
+		return err
+	}
+
+	if err = <-errCh; err != nil {
+		if errors.Is(err, ConcurrentDeleteTempTablesError) {
```

Comment:
This is a compile error for me. I think you want to import `github.com/cockroachdb/errors`?